### ISSUE TITLE
syncoid falls back to old snapshot due to the assumption that guids always increase over time

### ISSUE THREAD

**Issue by da-me** [jimsalterjrs/sanoid#112]
I noticed that syncing datasets to a secondary location takes much longer than usual. Further investigations showed that syncoid doesn't use the latest snapshot anymode, but instead falls back to one several days old. This results in much more transferred data, up to the point where the transfer doesn't finish in time for the next transfer.

I checked the syncoid source, and found this line: 

`foreach my $snap ( sort { $snaps{'source'}{$b}{'guid'}<=>$snaps{'source'}{$a}{'guid'} } keys %{ $snaps{'source'} }) {`

Syncoid seems to make the assumtion that the guid of a snapshots increases over time, so it sorts the list of snapshots by the guid - and using the one with the highest guid as the base for the transfer.

Unfortunately the guid isnt strictly increasing, for example:

```
zfs-pool0/rootfs@autosnap_2017-05-01_00:00:01_monthly   guid    16550336102538709723    -
zfs-pool0/rootfs@autosnap_2017-06-01_00:00:03_monthly   guid    15625873774715287149    -
zfs-pool0/rootfs@autosnap_2017-06-18_00:00:04_daily     guid    3406565320849952515     -
zfs-pool0/rootfs@autosnap_2017-06-19_00:00:01_daily     guid    13811447044692346659    -
zfs-pool0/rootfs@autosnap_2017-06-20_00:00:05_daily     guid    13832666677524663213    -
zfs-pool0/rootfs@autosnap_2017-06-21_00:00:01_daily     guid    12808608514254407328    -
zfs-pool0/rootfs@autosnap_2017-06-22_00:00:05_daily     guid    3675640657691258182     -
zfs-pool0/rootfs@autosnap_2017-06-23_00:00:05_daily     guid    2874086318737555719     -
zfs-pool0/rootfs@autosnap_2017-06-24_00:00:01_daily     guid    4946278501556411124     -
zfs-pool0/rootfs@autosnap_2017-06-25_00:00:01_daily     guid    12667147980110806263    -
zfs-pool0/rootfs@autosnap_2017-06-26_00:00:01_daily     guid    13290634381064213234    -
zfs-pool0/rootfs@autosnap_2017-06-27_00:00:04_daily     guid    2927481008951120106     -
zfs-pool0/rootfs@autosnap_2017-06-28_00:00:02_daily     guid    10102896560564605751    -
zfs-pool0/rootfs@autosnap_2017-06-29_00:00:04_daily     guid    3749114855812291004     -
zfs-pool0/rootfs@autosnap_2017-06-30_00:00:03_daily     guid    4913549755034086928     -
zfs-pool0/rootfs@autosnap_2017-07-01_00:00:02_daily     guid    17442182901804757611    -
zfs-pool0/rootfs@autosnap_2017-07-01_00:00:02_monthly   guid    1760519422107552411     -
zfs-pool0/rootfs@autosnap_2017-07-02_00:00:02_daily     guid    3811977750916262399     -
zfs-pool0/rootfs@autosnap_2017-07-03_00:00:22_daily     guid    14116272844453192459    -
zfs-pool0/rootfs@autosnap_2017-07-04_00:00:01_daily     guid    2432455564301095616     -
zfs-pool0/rootfs@autosnap_2017-07-05_00:00:01_daily     guid    2574631365735852057     -
zfs-pool0/rootfs@autosnap_2017-07-06_00:00:07_daily     guid    3491484778177975083     -
zfs-pool0/rootfs@autosnap_2017-07-07_00:00:02_daily     guid    10389491954339870304    -
zfs-pool0/rootfs@autosnap_2017-07-08_00:00:02_daily     guid    3967986398942558486     -
zfs-pool0/rootfs@autosnap_2017-07-09_00:00:05_daily     guid    16955241301340128739    -
zfs-pool0/rootfs@autosnap_2017-07-10_00:00:01_daily     guid    14499369380102693305    -
zfs-pool0/rootfs@autosnap_2017-07-11_00:00:02_daily     guid    14595055797110083951    -
zfs-pool0/rootfs@autosnap_2017-07-12_00:00:01_daily     guid    6758233741369754268     -
zfs-pool0/rootfs@autosnap_2017-07-13_00:00:01_daily     guid    8099885083852384996     -
zfs-pool0/rootfs@autosnap_2017-07-14_00:00:01_daily     guid    11261151948615802097    -
zfs-pool0/rootfs@autosnap_2017-07-15_00:00:07_daily     guid    4931791455786105256     -
zfs-pool0/rootfs@autosnap_2017-07-15_22:00:02_hourly    guid    4552930066328433015     -
zfs-pool0/rootfs@autosnap_2017-07-15_23:00:04_hourly    guid    14873157840306400114    -
zfs-pool0/rootfs@autosnap_2017-07-16_00:00:05_daily     guid    4691023830951142985     -
zfs-pool0/rootfs@autosnap_2017-07-16_00:00:05_hourly    guid    3077742037866020249     -
zfs-pool0/rootfs@autosnap_2017-07-16_01:00:04_hourly    guid    10412149894668565627    -
zfs-pool0/rootfs@autosnap_2017-07-16_02:00:05_hourly    guid    7701297432395582285     -
zfs-pool0/rootfs@autosnap_2017-07-16_03:00:05_hourly    guid    6695138346480861415     -
zfs-pool0/rootfs@autosnap_2017-07-16_04:00:04_hourly    guid    1121541588772947145     -
zfs-pool0/rootfs@autosnap_2017-07-16_05:00:05_hourly    guid    18053066738009427629    -
zfs-pool0/rootfs@autosnap_2017-07-16_06:00:05_hourly    guid    6935211421935821299     -
zfs-pool0/rootfs@autosnap_2017-07-16_07:00:05_hourly    guid    9325516668952621621     -
zfs-pool0/rootfs@autosnap_2017-07-16_08:00:04_hourly    guid    3202910617794217086     -
zfs-pool0/rootfs@autosnap_2017-07-16_09:00:04_hourly    guid    10064373481933678588    -
zfs-pool0/rootfs@autosnap_2017-07-16_10:00:04_hourly    guid    6595715464496989126     -
zfs-pool0/rootfs@autosnap_2017-07-16_11:00:04_hourly    guid    10819041052367391341    -
zfs-pool0/rootfs@autosnap_2017-07-16_12:00:01_hourly    guid    17085193039052289993    -
zfs-pool0/rootfs@autosnap_2017-07-16_13:00:01_hourly    guid    8450370091626383745     -
zfs-pool0/rootfs@autosnap_2017-07-16_14:00:02_hourly    guid    6798423817509036848     -
zfs-pool0/rootfs@autosnap_2017-07-16_15:00:01_hourly    guid    8167706024867539039     -
zfs-pool0/rootfs@autosnap_2017-07-16_16:00:02_hourly    guid    2549274257963201323     -
zfs-pool0/rootfs@autosnap_2017-07-16_17:00:01_hourly    guid    16865647181934894239    -
zfs-pool0/rootfs@autosnap_2017-07-16_18:00:01_hourly    guid    8560651667669058022     -
zfs-pool0/rootfs@autosnap_2017-07-16_19:00:01_hourly    guid    10880419460653065652    -
zfs-pool0/rootfs@autosnap_2017-07-16_20:00:01_hourly    guid    3739613450821695335     -
zfs-pool0/rootfs@autosnap_2017-07-16_21:00:01_hourly    guid    4485172514665231058     -
zfs-pool0/rootfs@autosnap_2017-07-16_22:00:02_hourly    guid    1810664790510384759     -
zfs-pool0/rootfs@autosnap_2017-07-16_23:00:02_hourly    guid    9543856695604516203     -
zfs-pool0/rootfs@autosnap_2017-07-17_00:00:02_hourly    guid    13657449188404604953    -
zfs-pool0/rootfs@autosnap_2017-07-17_00:00:02_daily     guid    15010589550929764921    -
zfs-pool0/rootfs@autosnap_2017-07-17_01:00:01_hourly    guid    39715379843673135       -
zfs-pool0/rootfs@autosnap_2017-07-17_02:00:01_hourly    guid    7474409435815234043     -
zfs-pool0/rootfs@autosnap_2017-07-17_03:00:01_hourly    guid    10762319413679482333    -
zfs-pool0/rootfs@autosnap_2017-07-17_04:00:01_hourly    guid    2583786760167441258     -
zfs-pool0/rootfs@autosnap_2017-07-17_05:00:01_hourly    guid    12525115470609591592    -
zfs-pool0/rootfs@autosnap_2017-07-17_06:00:02_hourly    guid    16601979248779388686    -
zfs-pool0/rootfs@autosnap_2017-07-17_07:00:01_hourly    guid    5850619516679431924     -
zfs-pool0/rootfs@autosnap_2017-07-17_08:00:02_hourly    guid    10968229830444476586    -
zfs-pool0/rootfs@autosnap_2017-07-17_09:00:02_hourly    guid    8777213751335149556     -
zfs-pool0/rootfs@test   guid    14765516649874425170    -
zfs-pool0/rootfs@autosnap_2017-07-17_10:00:02_hourly    guid    12775993295484572831    -
zfs-pool0/rootfs@syncoid_darkland_2017-07-17:10:01:03   guid    5676311719565828817     -
```

All snapshots except the one called "test" have been created with sanoid.

**Comment by jimsalterjrs** [comment]
Thank you @da-me! This was a significant regression; I accidentally updated ALL references to `creation` to `guid` when updating the codebase.  Originally I wrote strict snapshot matching to check `creation` time rather than `guid`; in a recent update I fixed that but BROKE sorting, as you discovered and reported.

Fixed in master, v1.4.14.

**Comment by da-me** [comment]
Looks good, just tested v1.4.14 - seems to be using the latest snapshot again. Thanks!
